common/src/safe_alloc.rs

```diff
@@ -0,0 +1,194 @@
+use std::ptr;
+
+pub use std::collections::TryReserveError;
+pub type TryReserveResult<T> = Result<T, TryReserveError>;
+
+pub trait VecExt: Sized {
+    type T;
+    fn try_with_capacity(n: usize) -> TryReserveResult<Self>;
+    fn try_push(&mut self, x: Self::T) -> TryReserveResult<()>;
+    fn try_extend_from_slice_copy(&mut self, other: &[Self::T]) -> TryReserveResult<()>
+    where
+        Self::T: Copy;
+    fn try_extend_from_slice_clone(&mut self, other: &[Self::T]) -> TryReserveResult<()>
+    where
+        Self::T: Clone;
+    fn try_append(&mut self, other: &mut Self) -> TryReserveResult<()>;
+}
+
+impl<T> VecExt for Vec<T> {
+    type T = T;
+    fn try_with_capacity(n: usize) -> TryReserveResult<Self> {
+        let mut v = Vec::new();
+        v.try_reserve_exact(n)?;
+        Ok(v)
+    }
+    fn try_push(&mut self, value: Self::T) -> TryReserveResult<()> {
+        let len = self.len();
+        if len == self.capacity() {
+            self.try_reserve(1)?;
+        }
+        unsafe {
+            let end = self.as_mut_ptr().add(len);
+            ptr::write(end, value);
+            self.set_len(len + 1);
+        }
+        Ok(())
+    }
+    fn try_extend_from_slice_copy(&mut self, other: &[T]) -> TryReserveResult<()>
+    where
+        T: Copy,
+    {
+        self.try_reserve(other.len())?;
+        unsafe { unchecked_append(self, other) };
+        Ok(())
+    }
+    fn try_extend_from_slice_clone(&mut self, other: &[Self::T]) -> TryReserveResult<()>
+    where
+        Self::T: Clone,
+    {
+        self.try_reserve(other.len())?;
+        let len = self.len();
+        unsafe {
+            let mut local_len = SetLenOnDrop::new(self);
+            let mut ptr = local_len.vec.as_mut_ptr().add(len);
+            for element in other {
+                ptr::write(ptr, element.clone());
+                ptr = ptr.offset(1);
+                local_len.increment_len(1);
+            }
+        }
+        Ok(())
+    }
+    fn try_append(&mut self, other: &mut Self) -> TryReserveResult<()> {
+        let other_len = other.len();
+        self.try_reserve(other_len)?;
+        unsafe {
+            other.set_len(0);
+            unchecked_append(self, std::slice::from_raw_parts(other.as_ptr(), other_len));
+        }
+        Ok(())
+    }
+}
+
+struct SetLenOnDrop<'a, T> {
+    vec: &'a mut Vec<T>,
+    local_len: usize,
+}
+
+impl<'a, T> SetLenOnDrop<'a, T> {
+    #[inline]
+    fn new(vec: &'a mut Vec<T>) -> Self {
+        SetLenOnDrop {
+            local_len: vec.len(),
+            vec,
+        }
+    }
+
+    #[inline]
+    fn increment_len(&mut self, increment: usize) {
+        self.local_len += increment;
+    }
+}
```

Comment:
This idea looks good. I think it can have similar interface like `MaybeUninit` because they have similar concept.

```rust
struct VecBuilder {
    vec: Vec<T>,
    len: usize,
}

impl VecUninit {
    fn try_with_len(len: usize) -> TryReserveResult {
        Self::try_with_capacity(len, len)
    }
    /// SAFETY: len <= capacity
    unsafe fn try_with_capacity(capacity: usize, len: usize) -> TryReserveResult {
        debug_assert!(len <= capacity);
        Self {
            vec: Vec::try_with_capacity(...)
            len,
        }
    }
    unsafe fn assume_init(self) -> Vec {
        self.vec.set_len(self.len);
        self.vec
    }
}

```

Reply:
Oh.. this is just copied from the standard library. I should add a comment.